rama-http/src/headers/x_robots_tag_components/robots_tag_components/builder.rs

```diff
@@ -0,0 +1,268 @@
+        self.0
+    }
+
+    pub fn add_custom_rule_simple(&mut self, key: HeaderValueString) -> &mut Self {
+        self.0.custom_rules.push(key.into());
+        self
+    }
+
+    pub fn add_custom_rule_composite(
+        &mut self,
+        key: HeaderValueString,
+        value: HeaderValueString,
+    ) -> &mut Self {
+        self.0.custom_rules.push((key, value).into());
+        self
+    }
+
+    pub fn set_unavailable_after(&mut self, unavailable_after: DateTime<Utc>) -> &mut Self {
+        self.0.unavailable_after = Some(unavailable_after.into());
+        self
+    }
+
+    pub fn unavailable_after(mut self, unavailable_after: DateTime<Utc>) -> Self {
+        self.0.unavailable_after = Some(unavailable_after.into());
+        self
+    }
+
+    robots_tag_builder_field!(bot_name, HeaderValueString, optional);
+    robots_tag_builder_field!(all, bool);
+    robots_tag_builder_field!(no_index, bool);
+    robots_tag_builder_field!(no_follow, bool);
+    robots_tag_builder_field!(none, bool);
+    robots_tag_builder_field!(no_snippet, bool);
+    robots_tag_builder_field!(index_if_embedded, bool);
+    robots_tag_builder_field!(max_snippet, u32);
+    robots_tag_builder_field!(max_image_preview, MaxImagePreviewSetting, optional);
+    robots_tag_builder_field!(max_video_preview, Option<u32>);
+    robots_tag_builder_field!(no_translate, bool);
+    robots_tag_builder_field!(no_image_index, bool);
+    robots_tag_builder_field!(no_ai, bool);
+    robots_tag_builder_field!(no_image_ai, bool);
+    robots_tag_builder_field!(spc, bool);
+
+    /// Adds a field based on its `&str` representation (also handles whitespace by trimming)
+    ///
+    /// # Returns and Errors
+    ///
+    /// - `Result<&mut Self, OpaqueError>`
+    ///     - `Ok(&mut Self)`
+    ///         - when the field was valid and successfully added
+    ///         - returns `&mut Self` wrapped inside for easier chaining of functions
+    ///     - `Err(OpaqueError)`
+    ///         - is of type [`headers::Error`] when the field name is not valid
+    ///         - for composite rules (key + value), wraps the conversion error for the value
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use std::num::ParseIntError;
+    /// # use rama_http::headers::x_robots_tag_components::RobotsTag;
+    /// let mut builder = RobotsTag::builder().no_follow();
+    ///
+    /// assert!(builder.add_field("nosnippet").is_ok());
+    /// assert!(builder.add_field("max-snippet: 8").is_ok());
+    /// assert!(builder.add_field("nonexistent").is_err_and(|e| e.is::<headers::Error>()));
+    /// assert!(builder.add_field("max-video-preview: not_a_number").is_err_and(|e| e.is::<ParseIntError>()));
+    ///
+    /// let robots_tag = builder.build();
+    ///
+    /// assert_eq!(robots_tag.no_snippet(), true);
+    /// assert_eq!(robots_tag.max_snippet(), 8);
+    /// ```
+    pub fn add_field(&mut self, s: &str) -> Result<&mut Self, OpaqueError> {
```

Comment:
why is this public? this is internal parser logic and I'm also still confused on how it like it is

Reply:
No best not, and I'm still also not convinced if that's the right abstraction even on a private level.

How I see it is that this is really just one big linear scan and whether something is a solo value, key-value pair or robot name is pretty clear from the context you find it in and the thing it is equal to.

Parsing I would definitely not expose outside of a TryFrom / FromStr / OtherTopLevelConstruct. We could if ever a good constructed feature requested is made for it of course, but not out of the box.